README.md
支持截获 `fetch`，其他用法和主仓库一样，增加的代码参见 [Support intercept fetch · zjffun/Mock-support-intercept-fetch@3a71000](https://github.com/zjffun/Mock-support-intercept-fetch/commit/3a71000aa125f79b4cd67cc971eab9574aaea834)

Support intercept `fetch`, other using are same as main repo.

使用：

```bash
npm install @zjffun/mockjs
```

- 2020-1-19: 支持截获请求的同时发送请求，方便在“控制台-网络”中查看具体请求信息。

使用：
```bash
// 设置模板前开启 sendRequest
Mock.sendRequest = true;
```

# Mock.js
<!-- 模拟请求 & 模拟数据 -->
[![Build Status](https://travis-ci.org/nuysoft/Mock.svg?branch=refactoring)](https://travis-ci.org/nuysoft/Mock)

<!-- [![Coverage Status](https://coveralls.io/repos/nuysoft/Mock/badge.png?branch=refactoring)](https://coveralls.io/r/nuysoft/Mock?branch=refactoring)
[![NPM version](https://badge.fury.io/js/mockjs.svg)](http://badge.fury.io/js/mockjs)
[![Bower version](https://badge.fury.io/bo/mockjs.svg)](http://badge.fury.io/bo/mockjs)
[![Dependency Status](https://gemnasium.com/nuysoft/Mock.svg)](https://gemnasium.com/nuysoft/Mock)
[![spm package](http://spmjs.io/badge/mockjs)](http://spmjs.io/package/mockjs) -->

Mock.js is a simulation data generator to help the front-end to develop and prototype separate from the back-end progress and reduce some monotony particularly while writing automated tests.

The official site: <http://mockjs.com>

## Features

* Generate simulated data according to the data template
* Provide request/response mocking for ajax requests
* ~~Generate simulated data according to HTML-based templates~~

This library is loosely inspired by Elijah Manor's post [Mocking
Introduction](http://www.elijahmanor.com/2013/04/angry-birds-of-javascript-green-bird.html), [mennovanslooten/mockJSON](https://github.com/mennovanslooten/mockJSON), [appendto/jquery-mockjax](https://github.com/appendto/jquery-mockjax) and [victorquinn/chancejs](https://github.com/victorquinn/chancejs/).

## Questions?
If you have any questions, please feel free to ask through [New Issue](https://github.com/nuysoft/Mock/issues/new).

## Reporting an Issue
Make sure the problem you're addressing is reproducible. Use <http://jsbin.com/> or <http://jsfiddle.net/> to provide a test page. Indicate what browsers the issue can be reproduced in. What version of Mock.js is the issue reproducible in. Is it reproducible after updating to the latest version?

## License
Mock.js is available under the terms of the [MIT License](./LICENSE).


CHANGELOG.md
# Changelog

## 2019.6.9 1.1.0

1. 支持 fetch

## 2015.12.03 V0.2.0-alpha2

1. 改用 webpack 打包

### 2014.3.20 V0.2.0-alpha1

1. 增加网站

### 2014.12.23 V0.2.0 重构代码

1. 改用 gulp 打包
2. 改用 mocha 重写测试用例
3. 改用 requirejs 重构代码

### 2014.6.24 V0.2.0 重构代码

1. 支持 UMD，包括：
    * 未打包前的代码
    * 打包后的代码
2. random CLI
    * --help 增加方法和参数说明
3. 重构文档站 @萝素
    * 增加《入门》
    * 单列《文档》
4. 测试用例
    * 重写测试用例
    * 同时支持 nodeunit 和 qunit
    * 同时支持 jQuery、KISSY、Zepto
    * 同时支持 KMD、AMD、CMD
5. 复写 XHR @行列 @霍庸
6. 废弃的功能
    * Mock.mockjax()
    * Mock.tpl()
    * Mock.xtpl()
7. Random.dateImage() 支持 node-canvas
8. Mock.valid(tpl, data)
9. Mock.toJOSNSchema()
10. Mock.mock(regexp) 
11. 完善地支持 node，代码中的：
    * window
    * document
    * XHRHttpRequest
12. 支持相对路径

### 2014.6.23 V0.1.5

1. [!] 修复 #28 #29，因为 jQuery 每个版本在 Ajax 实现上有些差异，导致在拦截 Ajax 请求时出现了兼容性问题（例如，方法 `xhr.onload()` 访问不到）。本次[测试](http://jsfiddle.net/8y8Fz/)并通过的 jQuery 版本有：

    * jQuery 2.1.0
    * jQuery 2.0.2
    * jQuery 1.11.0
    * jQuery 1.10.1
    * jQuery 1.9.1
    * jQuery 1.8.3
    * jQuery 1.7.2
    * jQuery 1.6.4

非常抱歉，这个问题一直困扰着 Mock.js 用户，在后面的版本中，会通过拦截 XMLHttpRequest 的方法“一劳永逸”地解决拦截 Ajax 的兼容和适配问题。

### 2014.6.18 V0.1.4

1. [!] 修复 #14 0.1.1版本试了好像jq1.10可以，1.11下$.ajax拦截没反应
2. [!] 修复 #22 异步加载js文件的时候发现问题
3. [!] 修复 #23 Mock.mockjax 导致 $.getScript 不执行回调
4. [!] 修复 #24 Window Firefox 30.0 引用 占位符 抛错
5. [!] 修复 #25 改变了非函数属性的顺序，查找起来不方便
6. [!] 修复 #26 生成规则 支持 负数 number|-100-+100
7. [!] 修复 #27 数据模板编辑器 格式化（Tidy） 时会丢掉 函数属性
8. [+] 数据模板编辑器 增加了 编辑区 和 生成结果区 的同步滚动
9. [!] test/nodeuinit > test/nodeunit

### 2014.5.26 V0.1.3

1. [!] 修复 #21

### 2014.5.26 V0.1.2

1. [!] 重构 Mock.mockjax()
2. [!] 更新 package.json/devDependencies
3. [+] 增加 懒懒交流会 PPT

### 2014.5.9 V0.1.2
1. [+] 支持 [`Mock.mock(rurl, rtype, template)`](http://mockjs.com/#mock)
2. [+] 支持 [`'name|min-max': {}`、`'name|count': {}`](http://mockjs.com/#语法规范)
3. [+] 支持 [`'name': function(){}`](http://mockjs.com/#语法规范)
4. [+] 新增占位符 [@NOW](http://mockjs.com/#now)
5. [+] 更新了 [语法规范](http://mockjs.com/#语法规范)

### 2013.9.6
1. 增加占位符 @DATAIMAGE
2. 解析占位符时**完全**忽略大小写

### 2013.9.3
1. 文档增加用法示例：Sea.js (CMD)、RequireJS (AMD)
2. 增加对 CMD 规范的支持
3. 生成 SourceMap 文件 `dist/mock-min.map`

### 2013.8.21
1. 100% 基于客户端模板生成模拟数据，支持 KISSY XTemplate。
1. 调整文件结构。

### 2013.8.11
1. 80% 基于客户端模板生成模拟数据。
1. 完善针对 KISSY XTemplate 的测试用例 [test/mock4tpl-xtpl-node.js](test/mock4tpl-xtpl-node.js)。
1. [Mock4Tpl](src/tpl/mock4tpl.js) 支持 Partials。
1. Mock 支持转义 @。
1. 更新 README.md，增加对 Mock4Tpl 的说明。
1. 完善 [demo](demo/)。
1. 减少 Mock、Mock4Tpl 暴漏的 API。

### 2013.8.7
1. 75% 基于客户端模板生成模拟数据。
1. 完善测试用例 [test/mock4tpl-node.js](test/mock4tpl-node.js)。
1. 重构文件和目录结构，把代码模块化。
1. 参考 Handlebars.js，引入 Jison 生成模板解析器。

#### 2013.8.2
1. 60% 基于客户端模板生成模拟数据。
1. 增加测试用例 [test/mock4tpl-node.js](test/mock4tpl-node.js)，参考自 <http://handlebarsjs.com/>。

#### 2013.7.31
1. 50% 基于客户端模板生成模拟数据。

#### 2013.7.18
1. 增加占位符 @COLOR。
1. 完善对占位符的解析，过滤掉 `#%&()?/.`。
1. 对“支持的占位符”分组。

#### 2013.7.12
1. Mock.mock(rurl, template) 的参数 rurl 可以是字符串或正则。
1. 把产生随机元数据的接口封装到 Mock.Random 中。
1. 增加对日期的格式化。
1. 增加占位符 @IMG、@PARAGRAPH、@SENTENCE、@WORD、@FIRST、@LAST、@NAME、@DOMAIN、@EMAIL、@IP、@ID。
1. 支持嵌套的占位符，例如 `@IMG(@AD_SIZE)`。
1. 支持把普通属性当作占位符使用，例如 `@IMG(@size)`。


test/test.mock.fetch.js
/* global console, require, chai, describe, before, it */
// fetch 请求（fetch HTTP Request）
var expect = chai.expect
var Mock, _

describe('Fetch', function() {
    before(function(done) {
        require(['mock', 'underscore'], function() {
            Mock = arguments[0]
            _ = arguments[1]
            expect(Mock).to.not.equal(undefined)
            expect(_).to.not.equal(undefined)
            done()
        })
    })

    function stringify(json) {
        return JSON.stringify(json /*, null, 4*/ )
    }

    describe('fetch()', function() {
        it('', function(done) {
            var that = this
            var url = Math.random()
            fetch(url)
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(404).to.be.equal(d.status)
                })
                .catch(e => console.log(e))
                .finally(_ => done())
        })
    })

    describe('Mock.mock( rurl, template )', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_template.json'

            Mock.mock(/rurl_template.json/, {
                'list|1-10': [{
                    'id|+1': 1,
                    'email': '@EMAIL'
                }]
            })

            Mock.setup({
                // timeout: 100,
                timeout: '10-50',
            })

            fetch(url)
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                })
                .catch(e => console.log(e))
                .finally(_ => done())
        })
    })

    describe('Mock.mock( rurl, function(options) )', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_function.json'

            Mock.mock(/rurl_function\.json/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url)
                expect(options.type).to.be.equal('GET')
                expect(options.body).to.be.equal(undefined)
                return Mock.mock({
                    'list|1-10': [{
                        'id|+1': 1,
                        'email': '@EMAIL'
                    }]
                })
            })

            fetch(url)
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                })
                .catch(e => console.log(e))
                .finally(_ => done())
        })
    })

    describe('Mock.mock( rurl, function(options) ) + GET + data', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_function.json'

            Mock.mock(/rurl_function\.json/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url + '?foo=1')
                expect(options.type).to.be.equal('GET')
                expect(options.body).to.be.equal(undefined)
                return Mock.mock({
                    'list|1-10': [{
                        'id|+1': 1,
                        'email': '@EMAIL'
                    }]
                })
            })

            fetch(url + '?foo=1')
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                })
                .catch(e => console.log(e))
                .finally(_ => done())
        })
    })

    describe('Mock.mock( rurl, function(options) ) + POST + data', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_function.json'

            Mock.mock(/rurl_function\.json/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url)
                expect(options.type).to.be.equal('POST')
                expect(options.body).to.deep.equal(stringify({ foo: 1 }))
                return Mock.mock({
                    'list|1-10': [{
                        'id|+1': 1,
                        'email': '@EMAIL'
                    }]
                })
            })

            fetch(url, {
                method: 'POST',
                body: stringify({ foo: 1 })
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                })
                .catch(e => console.log(e))
                .finally(_ => done())
        })
    })

    describe('Mock.mock( rurl, rtype, template )', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_rtype_template.json'
            var count = 0

            Mock.mock(/rurl_rtype_template\.json/, 'get', {
                'list|1-10': [{
                    'id|+1': 1,
                    'email': '@EMAIL',
                    type: 'get'
                }]
            })
            Mock.mock(/rurl_rtype_template\.json/, 'post', {
                'list|1-10': [{
                    'id|+1': 1,
                    'email': '@EMAIL',
                    type: 'post'
                }]
            })

            fetch(url, {
                method: 'GET'
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            fetch(url, {
                method: 'POST',
                body: stringify({ foo: 1 })
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += url + ' => ' + stringify(d)
                    expect(d).to.have.property('list')
                        .that.be.an('array').with.length.within(1, 10)
                    _.each(d.list, function(item, index, list) {
                        if (index > 0) expect(item.id).to.be.equal(list[index - 1].id + 1)
                    })
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            function success( /*data*/ ) {
                count++
            }

            function complete() {
                if (count === 2) done()
            }
        })
    })

    describe('Mock.mock( rurl, rtype, function(options) )', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_rtype_function.json'
            var count = 0

            Mock.mock(/rurl_rtype_function\.json/, /get/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url)
                expect(options.type).to.be.equal('GET')
                expect(options.body).to.be.equal(undefined)
                return {
                    type: options.type.toLowerCase()
                }
            })
            Mock.mock(/rurl_rtype_function\.json/, /post|put/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url)
                expect(['POST', 'PUT']).to.include(options.type)
                expect(options.body).to.be.equal(undefined)
                return {
                    type: options.type.toLowerCase()
                }
            })

            fetch(url, {
                method: 'GET'
            })
                .then(d => {
                    return d.json()
                })
                .then(d => {
                    that.test.title += 'GET ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'get')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            fetch(url, {
                method: 'POST'
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += 'POST ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'post')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            fetch(url, {
                method: 'PUT'
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += 'PUT ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'put')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            function success( /*data*/ ) {
                count++
            }

            function complete() {
                if (count === 3) done()
            }

        })
    })
    describe('Mock.mock( rurl, rtype, function(options) ) + data', function() {
        it('', function(done) {
            var that = this
            var url = 'rurl_rtype_function.json'
            var count = 0

            Mock.mock(/rurl_rtype_function\.json/, /get/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url + '?foo=1')
                expect(options.type).to.be.equal('GET')
                expect(options.body).to.be.equal(undefined)
                return {
                    type: options.type.toLowerCase()
                }
            })
            Mock.mock(/rurl_rtype_function\.json/, /post|put/, function(options) {
                expect(options).to.not.equal(undefined)
                expect(options.url).to.be.equal(url)
                expect(['POST', 'PUT']).to.include(options.type)
                expect(options.body).to.deep.equal(stringify({ foo: 1 }))
                return {
                    type: options.type.toLowerCase()
                }
            })

            fetch(url + '?foo=1', {
                method: 'GET'
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += 'GET ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'get')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            fetch(url, {
                method: 'POST',
                body: stringify({ foo: 1 })
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += 'POST ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'post')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            fetch(url, {
                method: 'PUT',
                body: stringify({ foo: 1 })
            })
                .then(d => d.json())
                .then(d => {
                    that.test.title += 'PUT ' + url + ' => ' + stringify(data)
                    expect(d).to.have.property('type', 'put')
                    success()
                })
                .catch(e => console.log(e))
                .finally(_ => complete())

            function success( /*data*/ ) {
                count++
            }

            function complete() {
                if (count === 3) done()
            }

        })
    })
})


src/mock/xhr/xhr.js
/* global window, document, location, Event, setTimeout */
/*
    ## MockXMLHttpRequest

    期望的功能：
    1. 完整地覆盖原生 XHR 的行为
    2. 完整地模拟原生 XHR 的行为
    3. 在发起请求时，自动检测是否需要拦截
    4. 如果不必拦截，则执行原生 XHR 的行为
    5. 如果需要拦截，则执行虚拟 XHR 的行为
    6. 兼容 XMLHttpRequest 和 ActiveXObject
        new window.XMLHttpRequest()
        new window.ActiveXObject("Microsoft.XMLHTTP")

    关键方法的逻辑：
    * new   此时尚无法确定是否需要拦截，所以创建原生 XHR 对象是必须的。
    * open  此时可以取到 URL，可以决定是否进行拦截。
    * send  此时已经确定了请求方式。

    规范：
    http://xhr.spec.whatwg.org/
    http://www.w3.org/TR/XMLHttpRequest2/

    参考实现：
    https://github.com/philikon/MockHttpRequest/blob/master/lib/mock.js
    https://github.com/trek/FakeXMLHttpRequest/blob/master/fake_xml_http_request.js
    https://github.com/ilinsky/xmlhttprequest/blob/master/XMLHttpRequest.js
    https://github.com/firebug/firebug-lite/blob/master/content/lite/xhr.js
    https://github.com/thx/RAP/blob/master/lab/rap.plugin.xinglie.js

    **需不需要全面重写 XMLHttpRequest？**
        http://xhr.spec.whatwg.org/#interface-xmlhttprequest
        关键属性 readyState、status、statusText、response、responseText、responseXML 是 readonly，所以，试图通过修改这些状态，来模拟响应是不可行的。
        因此，唯一的办法是模拟整个 XMLHttpRequest，就像 jQuery 对事件模型的封装。

    // Event handlers
    onloadstart         loadstart
    onprogress          progress
    onabort             abort
    onerror             error
    onload              load
    ontimeout           timeout
    onloadend           loadend
    onreadystatechange  readystatechange
 */

var Util = require('../util')

// 备份原生 XMLHttpRequest
window._XMLHttpRequest = window.XMLHttpRequest
window._ActiveXObject = window.ActiveXObject

/*
    PhantomJS
    TypeError: '[object EventConstructor]' is not a constructor (evaluating 'new Event("readystatechange")')

    https://github.com/bluerail/twitter-bootstrap-rails-confirm/issues/18
    https://github.com/ariya/phantomjs/issues/11289
*/
try {
    new window.Event('custom')
} catch (exception) {
    window.Event = function(type, bubbles, cancelable, detail) {
        var event = document.createEvent('CustomEvent') // MUST be 'CustomEvent'
        event.initCustomEvent(type, bubbles, cancelable, detail)
        return event
    }
}

var XHR_STATES = {
    // The object has been constructed.
    UNSENT: 0,
    // The open() method has been successfully invoked.
    OPENED: 1,
    // All redirects (if any) have been followed and all HTTP headers of the response have been received.
    HEADERS_RECEIVED: 2,
    // The response's body is being received.
    LOADING: 3,
    // The data transfer has been completed or something went wrong during the transfer (e.g. infinite redirects).
    DONE: 4
}

var XHR_EVENTS = 'readystatechange loadstart progress abort error load timeout loadend'.split(' ')
var XHR_REQUEST_PROPERTIES = 'timeout withCredentials'.split(' ')
var XHR_RESPONSE_PROPERTIES = 'readyState responseURL status statusText responseType response responseText responseXML'.split(' ')

// https://github.com/trek/FakeXMLHttpRequest/blob/master/fake_xml_http_request.js#L32
var HTTP_STATUS_CODES = {
    100: "Continue",
    101: "Switching Protocols",
    200: "OK",
    201: "Created",
    202: "Accepted",
    203: "Non-Authoritative Information",
    204: "No Content",
    205: "Reset Content",
    206: "Partial Content",
    300: "Multiple Choice",
    301: "Moved Permanently",
    302: "Found",
    303: "See Other",
    304: "Not Modified",
    305: "Use Proxy",
    307: "Temporary Redirect",
    400: "Bad Request",
    401: "Unauthorized",
    402: "Payment Required",
    403: "Forbidden",
    404: "Not Found",
    405: "Method Not Allowed",
    406: "Not Acceptable",
    407: "Proxy Authentication Required",
    408: "Request Timeout",
    409: "Conflict",
    410: "Gone",
    411: "Length Required",
    412: "Precondition Failed",
    413: "Request Entity Too Large",
    414: "Request-URI Too Long",
    415: "Unsupported Media Type",
    416: "Requested Range Not Satisfiable",
    417: "Expectation Failed",
    422: "Unprocessable Entity",
    500: "Internal Server Error",
    501: "Not Implemented",
    502: "Bad Gateway",
    503: "Service Unavailable",
    504: "Gateway Timeout",
    505: "HTTP Version Not Supported"
}

/*
    MockXMLHttpRequest
*/

function MockXMLHttpRequest() {
    // 初始化 custom 对象，用于存储自定义属性
    this.custom = {
        events: {},
        requestHeaders: {},
        responseHeaders: {}
    }
}

MockXMLHttpRequest._settings = {
    timeout: '10-100',
    /*
        timeout: 50,
        timeout: '10-100',
     */
}

MockXMLHttpRequest.setup = function(settings) {
    Util.extend(MockXMLHttpRequest._settings, settings)
    return MockXMLHttpRequest._settings
}

Util.extend(MockXMLHttpRequest, XHR_STATES)
Util.extend(MockXMLHttpRequest.prototype, XHR_STATES)

// 标记当前对象为 MockXMLHttpRequest
MockXMLHttpRequest.prototype.mock = true

// 是否拦截 Ajax 请求
MockXMLHttpRequest.prototype.match = false

// 初始化 Request 相关的属性和方法
Util.extend(MockXMLHttpRequest.prototype, {
    // https://xhr.spec.whatwg.org/#the-open()-method
    // Sets the request method, request URL, and synchronous flag.
    open: function(method, url, async, username, password) {
        var that = this

        Util.extend(this.custom, {
            method: method,
            url: url,
            async: typeof async === 'boolean' ? async : true,
            username: username,
            password: password,
            options: {
                url: url,
                type: method
            }
        })

        this.custom.timeout = function(timeout) {
            if (typeof timeout === 'number') return timeout
            if (typeof timeout === 'string' && !~timeout.indexOf('-')) return parseInt(timeout, 10)
            if (typeof timeout === 'string' && ~timeout.indexOf('-')) {
                var tmp = timeout.split('-')
                var min = parseInt(tmp[0], 10)
                var max = parseInt(tmp[1], 10)
                return Math.round(Math.random() * (max - min)) + min
            }
        }(MockXMLHttpRequest._settings.timeout)

        // 查找与请求参数匹配的数据模板
        var item = find(this.custom.options)

        function handle(event) {
            // 同步属性 NativeXMLHttpRequest => MockXMLHttpRequest
            for (var i = 0; i < XHR_RESPONSE_PROPERTIES.length; i++) {
                try {
                    that[XHR_RESPONSE_PROPERTIES[i]] = xhr[XHR_RESPONSE_PROPERTIES[i]]
                } catch (e) {}
            }
            // 触发 MockXMLHttpRequest 上的同名事件
            that.dispatchEvent(new Event(event.type /*, false, false, that*/ ))
        }

        // 如果未找到匹配的数据模板，则采用原生 XHR 发送请求。
        if (!item) {
            // 创建原生 XHR 对象，调用原生 open()，监听所有原生事件
            var xhr = createNativeXMLHttpRequest()
            this.custom.xhr = xhr

            // 初始化所有事件，用于监听原生 XHR 对象的事件
            for (var i = 0; i < XHR_EVENTS.length; i++) {
                xhr.addEventListener(XHR_EVENTS[i], handle)
            }

            // xhr.open()
            if (username) xhr.open(method, url, async, username, password)
            else xhr.open(method, url, async)

            // 同步属性 MockXMLHttpRequest => NativeXMLHttpRequest
            for (var j = 0; j < XHR_REQUEST_PROPERTIES.length; j++) {
                try {
                    xhr[XHR_REQUEST_PROPERTIES[j]] = that[XHR_REQUEST_PROPERTIES[j]]
                } catch (e) {}
            }

            return
        }

        // 找到了匹配的数据模板，开始拦截 XHR 请求
        // sendRequest 为 true，拦截同时发送请求
        if(item.sendRequest){
            // 创建原生 XHR 对象，调用原生 open()，监听所有原生事件
            var xhr = createNativeXMLHttpRequest()
            this.custom.xhr = xhr

            // xhr.open()
            if (username) xhr.open(method, url, async, username, password)
            else xhr.open(method, url, async)

            // 同步属性 MockXMLHttpRequest => NativeXMLHttpRequest
            for (var j = 0; j < XHR_REQUEST_PROPERTIES.length; j++) {
                try {
                    xhr[XHR_REQUEST_PROPERTIES[j]] = that[XHR_REQUEST_PROPERTIES[j]]
                } catch (e) {}
            }
        }

        this.match = true
        this.custom.template = item
        this.readyState = MockXMLHttpRequest.OPENED
        this.dispatchEvent(new Event('readystatechange' /*, false, false, this*/ ))
    },
    // https://xhr.spec.whatwg.org/#the-setrequestheader()-method
    // Combines a header in author request headers.
    setRequestHeader: function(name, value) {
        // 原生 XHR
        if (!this.match) {
            this.custom.xhr.setRequestHeader(name, value)
            return
        }

        // 拦截 XHR
        var requestHeaders = this.custom.requestHeaders
        if (requestHeaders[name]) requestHeaders[name] += ',' + value
        else requestHeaders[name] = value
    },
    timeout: 0,
    withCredentials: false,
    upload: {},
    // https://xhr.spec.whatwg.org/#the-send()-method
    // Initiates the request.
    send: function send(data) {
        var that = this
        this.custom.options.body = data

        // 原生 XHR
        if (!this.match) {
            this.custom.xhr.send(data)
            return
        }

        // 拦截 XHR
        // 拦截同时发送请求
        if(this.custom.xhr){
            this.custom.xhr.send(data)
        }

        // X-Requested-With header
        this.setRequestHeader('X-Requested-With', 'MockXMLHttpRequest')

        // loadstart The fetch initiates.
        this.dispatchEvent(new Event('loadstart' /*, false, false, this*/ ))

        if (this.custom.async) setTimeout(done, this.custom.timeout) // 异步
        else done() // 同步

        function done() {
            that.readyState = MockXMLHttpRequest.HEADERS_RECEIVED
            that.dispatchEvent(new Event('readystatechange' /*, false, false, that*/ ))
            that.readyState = MockXMLHttpRequest.LOADING
            that.dispatchEvent(new Event('readystatechange' /*, false, false, that*/ ))

            that.status = 200
            that.statusText = HTTP_STATUS_CODES[200]

            // fix #92 #93 by @qddegtya
            that.response = that.responseText = JSON.stringify(
                convert(that.custom.template, that.custom.options),
                null, 4
            )

            that.readyState = MockXMLHttpRequest.DONE
            that.dispatchEvent(new Event('readystatechange' /*, false, false, that*/ ))
            that.dispatchEvent(new Event('load' /*, false, false, that*/ ));
            that.dispatchEvent(new Event('loadend' /*, false, false, that*/ ));
        }
    },
    // https://xhr.spec.whatwg.org/#the-abort()-method
    // Cancels any network activity.
    abort: function abort() {
        // 原生 XHR
        if (!this.match) {
            this.custom.xhr.abort()
            return
        }

        // 拦截 XHR
        this.readyState = MockXMLHttpRequest.UNSENT
        this.dispatchEvent(new Event('abort', false, false, this))
        this.dispatchEvent(new Event('error', false, false, this))
    }
})

// 初始化 Response 相关的属性和方法
Util.extend(MockXMLHttpRequest.prototype, {
    responseURL: '',
    status: MockXMLHttpRequest.UNSENT,
    statusText: '',
    // https://xhr.spec.whatwg.org/#the-getresponseheader()-method
    getResponseHeader: function(name) {
        // 原生 XHR
        if (!this.match) {
            return this.custom.xhr.getResponseHeader(name)
        }

        // 拦截 XHR
        return this.custom.responseHeaders[name.toLowerCase()]
    },
    // https://xhr.spec.whatwg.org/#the-getallresponseheaders()-method
    // http://www.utf8-chartable.de/
    getAllResponseHeaders: function() {
        // 原生 XHR
        if (!this.match) {
            return this.custom.xhr.getAllResponseHeaders()
        }

        // 拦截 XHR
        var responseHeaders = this.custom.responseHeaders
        var headers = ''
        for (var h in responseHeaders) {
            if (!responseHeaders.hasOwnProperty(h)) continue
            headers += h + ': ' + responseHeaders[h] + '\r\n'
        }
        return headers
    },
    overrideMimeType: function( /*mime*/ ) {},
    responseType: '', // '', 'text', 'arraybuffer', 'blob', 'document', 'json'
    response: null,
    responseText: '',
    responseXML: null
})

// EventTarget
Util.extend(MockXMLHttpRequest.prototype, {
    addEventListener: function addEventListener(type, handle) {
        var events = this.custom.events
        if (!events[type]) events[type] = []
        events[type].push(handle)
    },
    removeEventListener: function removeEventListener(type, handle) {
        var handles = this.custom.events[type] || []
        for (var i = 0; i < handles.length; i++) {
            if (handles[i] === handle) {
                handles.splice(i--, 1)
            }
        }
    },
    dispatchEvent: function dispatchEvent(event) {
        var handles = this.custom.events[event.type] || []
        for (var i = 0; i < handles.length; i++) {
            handles[i].call(this, event)
        }

        var ontype = 'on' + event.type
        if (this[ontype]) this[ontype](event)
    }
})

// Inspired by jQuery
function createNativeXMLHttpRequest() {
    var isLocal = function() {
        var rlocalProtocol = /^(?:about|app|app-storage|.+-extension|file|res|widget):$/
        var rurl = /^([\w.+-]+:)(?:\/\/([^\/?#:]*)(?::(\d+)|)|)/
        var ajaxLocation = location.href
        var ajaxLocParts = rurl.exec(ajaxLocation.toLowerCase()) || []
        return rlocalProtocol.test(ajaxLocParts[1])
    }()

    return window.ActiveXObject ?
        (!isLocal && createStandardXHR() || createActiveXHR()) : createStandardXHR()

    function createStandardXHR() {
        try {
            return new window._XMLHttpRequest();
        } catch (e) {}
    }

    function createActiveXHR() {
        try {
            return new window._ActiveXObject("Microsoft.XMLHTTP");
        } catch (e) {}
    }
}


// 查找与请求参数匹配的数据模板：URL，Type
function find(options) {

    for (var sUrlType in MockXMLHttpRequest.Mock._mocked) {
        var item = MockXMLHttpRequest.Mock._mocked[sUrlType]
        if (
            (!item.rurl || match(item.rurl, options.url)) &&
            (!item.rtype || match(item.rtype, options.type.toLowerCase()))
        ) {
            // console.log('[mock]', options.url, '>', item.rurl)
            return item
        }
    }

    function match(expected, actual) {
        if (Util.type(expected) === 'string') {
            return expected === actual
        }
        if (Util.type(expected) === 'regexp') {
            return expected.test(actual)
        }
    }

}

// 数据模板 ＝> 响应数据
function convert(item, options) {
    return Util.isFunction(item.template) ?
        item.template(options) : MockXMLHttpRequest.Mock.mock(item.template)
}

module.exports = MockXMLHttpRequest
